planner/optimize.go

```diff
@@ -118,109 +119,132 @@ func Optimize(ctx context.Context, sctx sessionctx.Context, node ast.Node, is in
+	var names types.NameSlice
+	var bestPlan, bestPlanFromBind plannercore.Plan
+	if useBinding {
+		minCost := math.MaxFloat64
+		var bindStmtHints stmtctx.StmtHints
+		originHints := hint.CollectHint(stmtNode)
+		// bindRecord must be not nil when coming here, try to find the best binding.
+		for _, binding := range bindRecord.Bindings {
+			if binding.Status != bindinfo.Using {
+				continue
+			}
+			metrics.BindUsageCounter.WithLabelValues(scope).Inc()
+			hint.BindHint(stmtNode, binding.Hint)
+			curStmtHints, _, curWarns := handleStmtHints(binding.Hint.GetFirstTableHints())
+			sessVars.StmtCtx.StmtHints = curStmtHints
+			plan, curNames, cost, err := optimize(ctx, sctx, node, is)
+			if err != nil {
+				binding.Status = bindinfo.Invalid
+				handleInvalidBindRecord(ctx, sctx, scope, bindinfo.BindRecord{
+					OriginalSQL: bindRecord.OriginalSQL,
+					Db:          bindRecord.Db,
+					Bindings:    []bindinfo.Binding{binding},
+				})
+				continue
+			}
+			if cost < minCost {
+				bindStmtHints, warns, minCost, names, bestPlanFromBind = curStmtHints, curWarns, cost, curNames, plan
+			}
 		}
-		return bestPlan, names, nil
-	}
-	bestCostAmongHints := math.MaxFloat64
-	var bestPlanAmongHints plannercore.Plan
-	originHints := hint.CollectHint(stmtNode)
-	// Try to find the best binding.
-	for _, binding := range bindRecord.Bindings {
-		if binding.Status != bindinfo.Using {
-			continue
+		if bestPlanFromBind == nil {
+			sessVars.StmtCtx.AppendWarning(errors.New("no plan generated from bindings"))
+		} else {
+			bestPlan = bestPlanFromBind
+			sessVars.StmtCtx.StmtHints = bindStmtHints
+			for _, warn := range warns {
+				sessVars.StmtCtx.AppendWarning(warn)
+			}
+			if err := setFoundInBinding(sctx, true); err != nil {
+				logutil.BgLogger().Warn("set tidb_found_in_binding failed", zap.Error(err))
+			}
 		}
-		metrics.BindUsageCounter.WithLabelValues(scope).Inc()
-		hint.BindHint(stmtNode, binding.Hint)
-		curStmtHints, curWarns := handleStmtHints(binding.Hint.GetFirstTableHints())
-		sctx.GetSessionVars().StmtCtx.StmtHints = curStmtHints
-		plan, _, cost, err := optimize(ctx, sctx, node, is)
+		// Restore the hint to avoid changing the stmt node.
+		hint.BindHint(stmtNode, originHints)
+	}
+	// No plan found from the bindings, or the bindings are ignored.
+	if bestPlan == nil {
+		sessVars.StmtCtx.StmtHints = originStmtHints
+		bestPlan, names, _, err = optimize(ctx, sctx, node, is)
 		if err != nil {
-			binding.Status = bindinfo.Invalid
-			handleInvalidBindRecord(ctx, sctx, scope, bindinfo.BindRecord{
-				OriginalSQL: bindRecord.OriginalSQL,
-				Db:          bindRecord.Db,
-				Bindings:    []bindinfo.Binding{binding},
-			})
-			continue
+			return nil, nil, err
 		}
-		if cost < bestCostAmongHints {
-			if sctx.GetSessionVars().UsePlanBaselines {
-				stmtHints, warns = curStmtHints, curWarns
+	}
+
+	// Add a baseline evolution task if:
+	// 1. the returned plan is from bindings;
```

Comment:
Why do we add this condition? If the bestPlan is not from bindings, why we can not add a baseline evolution task?

Reply:
If the bestPlan is not from bindings, it means that there is no binding available for this query, or the binding is problematic and fails to generate a plan, so do not add evolution task for it.